common/util/trees.h

```diff
@@ -0,0 +1,856 @@
+  // child nodes. It must be move-assignable to, and move-constructible from,
+  // a value returned by `Children()` method.
+  using container_type = detected_or_t<std::remove_reference_t<ChildrenType_>,
+                                       TreeNodeSubnodesType, Node>;
+};
+
+// Defined when `Node` contains `Value()` method.
+template <typename Node,  //
+          typename Value_ = decltype(std::declval<Node>().Value())>
+struct TreeNodeValueTraits : FeatureTraits {
+  // Type of data returned by `Value()` method.
+  using type = std::remove_reference_t<Value_>;
+  // Aliases provided for convenience
+  using reference = std::add_lvalue_reference_t<type>;
+  using const_reference = std::add_lvalue_reference_t<const type>;
+};
+
+// Defined when `Node` contains `Parent()` method which dereferences to a type
+// fulfilling NodeTraits concept.
+template <typename Node,  //
+          typename Parent_ = decltype(*std::declval<Node>().Parent()),
+          typename = std::void_t<TreeNodeChildrenTraits<Parent_>>>
+struct TreeNodeParentTraits : FeatureTraits {};
+
+// BirthRank implementation details:
+
+// BirthRank implementation supportin any container
+template <class T>
+inline static size_t BirthRank(const T& node, std::input_iterator_tag) {
+  if (node.Parent() != nullptr) {
+    size_t index = 0;
+    for (const auto& child : node.Parent()->Children()) {
+      if (&node == &child) return index;
+      ++index;
+    }
+  }
+  return 0;
+}
+
+// BirthRank optimized for random access containers
+template <class T>
+inline static size_t BirthRank(const T& node, std::random_access_iterator_tag) {
+  if (node.Parent() != nullptr) {
+    return std::distance(&(node.Parent()->Children().front()), &node);
+  }
+  return 0;
+}
+
+}  // namespace trees_internal
+
+// TreeNodeTraits:
+
+// Traits of a type fulfilling TreeNode concept.
+// `TreeNodeTraits<T>` is defined for every class `T` fulfilling the TreeNode
+// concept. It can be used in SFINAE tests.
+template <class Node,  //
+          typename Children_ = trees_internal::TreeNodeChildrenTraits<Node>>
+struct TreeNodeTraits : trees_internal::FeatureTraits {
+  using Parent = detected_or_t<trees_internal::UnavailableFeatureTraits,
+                               trees_internal::TreeNodeParentTraits, Node>;
+  using Value = detected_or_t<trees_internal::UnavailableFeatureTraits,
+                              trees_internal::TreeNodeValueTraits, Node>;
+  using Children = Children_;
+};
+
+// Functions operating on tree nodes:
+
+// Returns true when `node` is a leaf, false otherwise.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::available>* = nullptr>
+bool is_leaf(T& node) {
+  return node.Children().empty();
+}
+
+// Descend through children using indices specified by iterator range.
+// Iterator type must dereference to an integral value.
+// This works on any internal node, not just the root.
+template <class T, class Iterator,  //
```

Comment:
// This implementation fatally exits if any indices are out-of-range.
// The const-ness of the returned reference matches the const-ness of the node argumnent.

[The const-ness comment is good to repeat where applicable.]

Reply:
Added.
Mentioned in all (I hope so) relevant functions.